examples/udp_client/linux_udp_client.cc

```diff
@@ -0,0 +1,136 @@
+#include <unistd.h> // For close()
+#include <fstream>
+#include <iostream>
+
+#include <point_one/fusion_engine/parsers/fusion_engine_framer.h>
+
+#include "../common/print_message.h"
+
+#define DEBUG_ON 0
+
+using namespace point_one::fusion_engine::examples;
+using namespace point_one::fusion_engine::parsers;
+
+static bool shutdown_pending_ = false;
+
+/******************************************************************************/
+void HandleSignal(int signal) {
+  if (signal == SIGINT || signal == SIGTERM) {
+    std::signal(signal, SIG_DFL);
+    shutdown_pending_ = true;
+  }
+}
+
+/******************************************************************************/
+void* GetInAddr(struct sockaddr* sa) {
+  if (sa->sa_family == AF_INET) {
+    return &(((struct sockaddr_in*)sa)->sin_addr);
+  }
+  return &(((struct sockaddr_in6*)sa)->sin6_addr);
+}
+
+/******************************************************************************/
+int main(int argc, const char* argv[]) {
+  // Parse arguments.
+  if (argc > 2) {
+    printf(R"EOF(
+Usage: %s [PORT]
+
+Receive incoming data from an Atlas device over UDP and print out the incoming message
+contents.
+)EOF",
+           argv[0]);
+    return 0;
+  }
+
+  int port = (argc == 2) ? std::stoi(argv[1]) : 12345;
+
+  // Create UDP socket.
+  int sock = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+  if (sock < 0) {
+    printf("Error creating socket.\n");
+    return 2;
+  }
+
+  // Bind socket to port.
+  sockaddr_in addr;
+  addr.sin_family = AF_INET;
+  addr.sin_port = htons(port);
+  addr.sin_addr.s_addr = INADDR_ANY; // Any local address.
+  int ret = bind(sock, (struct sockaddr*)&addr, sizeof(addr));
+  if (ret < 0) {
+    close(sock);
+    printf("Error binding.\n");
+    return 3;
+  }
+
+  // Listen for SIGINT (Ctrl-C) or SIGTERM and shutdown gracefully.
+  std::signal(SIGINT, HandleSignal);
+  std::signal(SIGTERM, HandleSignal);
+
+  // Receive incoming data.
+  FusionEngineFramer framer(1024);
+  framer.SetMessageCallback(PrintMessage);
+
+  uint8_t buffer[1024];
+  size_t total_bytes_read = 0;
+  struct sockaddr_storage their_addr; // Address of recieved packet.
+  socklen_t addr_len = sizeof(their_addr);
+  ret = 0;
+  char their_ip[INET6_ADDRSTRLEN];
+  while (!shutdown_pending_) {
+    ssize_t bytes_read = recvfrom(sock, buffer, sizeof(buffer), 0,
+                                  (struct sockaddr*)&their_addr, &addr_len);
+
+    if (bytes_read < 0) {
+      printf("Error reading from socket: %s (%d)\n", std::strerror(errno),
+             errno);
+      ret = 4;
+      break;
+    } else if (bytes_read == 0) {
+      printf("Socket closed remotely.\n");
+      break;
+    }
+
+    inet_ntop(their_addr.ss_family, GetInAddr((struct sockaddr*)&their_addr),
+              their_ip, sizeof(their_ip));
+    buffer[bytes_read] = '\0';
+    if (DEBUG_ON) {
+      // Enable to debug udp server to client problems.
+      printf("listener: received raw packet [%s] from %s\n", buffer, their_ip);
```

Comment:
If we are keeping this print, it is very important _not_ to print out `buffer` as a string. It's ok to print `bytes_read` perhaps, but `buffer` contains arbitrary binary data and interpreting it as a string can have unintended consequences.

Reply:
done.